server/src/main/java/com/cloud/vm/UserVmManagerImpl.java

```diff
@@ -4451,18 +4460,14 @@ public void finalizeStop(VirtualMachineProfile profile, Answer answer) {
             throw new PermissionDeniedException("The owner of " + vm + " is disabled: " + vm.getAccountId());
         }
 
-        Host destinationHost = null;
-        if (hostId != null) {
-            Account account = CallContext.current().getCallingAccount();
-            if (!_accountService.isRootAdmin(account.getId())) {
-                throw new PermissionDeniedException(
-                        "Parameter hostid can only be specified by a Root Admin, permission denied");
-            }
-            destinationHost = _hostDao.findById(hostId);
-            if (destinationHost == null) {
-                throw new InvalidParameterValueException("Unable to find the host to deploy the VM, host id=" + hostId);
-            }
-        }
+        Account account = CallContext.current().getCallingAccount();
+        boolean isRootAdmin = _accountService.isRootAdmin(account.getId());
+
+        Pod destinationPod = getDestinationPod(podId, isRootAdmin);
+        Cluster destinationCluster = getDestinationCluster(clusterId, isRootAdmin);
+        Host destinationHost = getDestinationHost(hostId, isRootAdmin);
```

Comment:
and next passes a host that is not in the passed cluster?

Reply:
I have simplified this to by using an if-else if block that checks from the lowest unit (host), followed cluster if host is not specified, and lastly pod.
`if (host != null){
    planner = host planner

} else if(cluster != null) {
    planner = cluster planner

}else if(pod != null)  {
   planner = pod planner
}
`

Removed the validation that checks the parent-child relations between pod, cluster and host